### ISSUE TITLE
Allow each component to be built into an independent package

### ISSUE THREAD

**Issue by trevor-vaughan** [puppetlabs/community-enablement-pipeline#6]
The AIO concept is fine but it would probably be easier for community involvement to provide something like rake tasks that call `fpm` and "do the right thing" for each component.

This would also allow easy integration into EPEL and other services that require clean build stacks.

**Comment by Sharpie** [comment]
AIO provides two main user benefits:

  - "one file" to copy to an agent node in order to get an installation rolling.
  - the ability to bring modern components to Long Term Support platforms that are determined to live forever (this also pays a large benefit on the development side by reducing the spread of versions in the test matrix).

The first point is weakening a bit as new platforms, notably SLES 15, have started using the system repos for heavy components like Boost. This is an approach I followed to get debian-9-armhf builds going, but the tradeoff is that you have to configure apt/yum/zypper in order to perform dependency resolution instead of just pointing dpkg or rpm at a single file.

The second point is also becoming less relevant now that the frenetic days of changes between Ruby 1.8 -> 1.9 -> 2.0 are behind us as Ruby 2.x has been fairly stable for a few years with few breaking changes.

That said, this would be a large project in its self and something that should be looked as separate  from "getting `puppet-agent` builds happening outside the Puppet Inc. firewall".

**Comment by trevor-vaughan** [comment]
The issue is that the AIO mentality breaks the ability to update and fix independent components.

There is certainly no issue with *namespacing* all of the puppet items under /opt/puppet and even having an interdependent set of items. However, the AIO approach prevents users from being able to fix and modify individual components without learning the internals of a complex and foreign build system.

A simple `fpm` wrapped in a script is generally understandable by everyone as is a simple wrapper that goes and prods each component in the correct order. Most importantly, it opens the gates to allowing larger community maintenance of packages in SLES, EPEL, etc... repos instead of relying on Puppet personnel to figure it all out and decide to do it.

**Comment by trevor-vaughan** [comment]
Some references from days gone by:

* [Puppet 2->3 was easier than 3->4 because AIO](https://groups.google.com/d/msg/puppet-users/goOeNjELMz8/WL6nghTLBgAJ)
* [The build scripts have to work, and always work](https://groups.google.com/d/msg/puppet-users/lIrm7CfZjNk/fUjbFGfQkHwJ)
* [Testing patches is insanely difficult](https://groups.google.com/d/topic/puppet-dev/L8tlJ3bWfIo/discussion)
* [No clue how to easily test on newer/OS versions of Ruby](https://groups.google.com/d/msg/puppet-dev/InN9aB0X_yI/xeRdPBBbBgAJ)
* [Trying to get new versions of things into newer OSs than Puppet officially supports is hard/impossible](https://groups.google.com/d/topic/puppet-dev/WseDYN-vIJo/discussion)
* [AIO increases vendor CVE exposure and maintenance burden leaving no simple recourse for end users but "wait on the vendor"](https://groups.google.com/d/msg/puppet-dev/qZ-nOvmfrig/P9jK_HL5XYEJ)

Honestly, in a search for 'AIO' on both `puppet-users` and `puppet-dev` I didn't find a single community member that had anything good to say about the approach (mostly it was just having no idea how anything worked). It was, of course, noted that it was now "easy to install" but at the detriment of a community being as the mercy of Puppet, Inc. having time to fix things and package them up for them or to start rolling the entire thing themselves.

**Comment by Sharpie** [comment]
> The issue is that the AIO mentality breaks the ability to update and fix independent components.

I don't agree with that. Individual components can still be fixed, but the delivery mechanism for the fixes is a new `puppet-agent` package release. For bug fixes or security patches this is a .z release.

The AIO mentality does mean trading off the fine-grained control of updating individual components directly. However, that does result in a much simpler test matrix. A given tagged set of components either works together, or it doesn't.

**Comment by trevor-vaughan** [comment]
Sure but *as a user, contributor, and developer* either your build magic has to work 100% of the time or you need new magic which is the less fun summary of the mailing list posts. TL;DR; we all gave up.

**Comment by Sharpie** [comment]
I absolutely agree with that. This isn't the first, second, or even third time I've attempted to build a complete `puppet-agent` package --- I've given up multiple times.

At the moment I want to make things less magical, better documented, and most importantly decoupled from "black boxes" that live behind the Puppet Inc. firewall. Splitting things up and moving them to another build system like `fpm` is something to consider. However, it's a whole other elephant that I don't have the bandwidth to focus on eating at the moment.

It is an elephant that we should talk about though :)

**Comment by trevor-vaughan** [comment]
So, after actually figuring out `vanagon` and doing a full AIO build, it turns out that each individual component already has individual build scripts in place that are 100% super `fpm` friendly. This is probably going to be faster than pretty much anything else you can do with the possible exception of publishing native versions of the puppet-build-tools and a bunch of working SRPMs (for EL systems anyway).

`vanagon` can output SRPMs, it just doesn't by default. So, I guess, do you want to fix `vanagon`, or do you want to start decoupling all the parts? I'm not sure what the third option is.

**Comment by jcbollinger** [comment]
I was pretty negative about the AIO idea when it was introduced, and I was vocal about it on Puppet social media channels.  My opinion has not changed much over the years -- I still dislike the AIO, but not enough to expend the effort to break it up for myself.

I would like to have the option to use the OS's or my own packages to satisfy Puppet's dependencies, instead of vendored packages in an AIO, and Trevor's references to past discussions cover most of my reasons.  On the other hand, I am sympathetic to Puppet's desire to have control of the software stack, yet I think that those concerns could be addressed as well or better by Puppet providing their vendored components as separate packages, instead of all bundled into a monolithic AIO.  Such a breakup would serve a few of my own ends as well, and would facilitate me taking care of many of the rest myself.

**Comment by Dan33l** [comment]
I was pretty surprised and perhaps also negative about the AIO idea when it was introduced.
But, after the `PC1` passed with this strange naming, the current naming is quite clean. And so AIO permits to have a matrix of compatibility simple and clear, even for module developers 

At Voxpupuli, very often, when we switch from strings facts like `$fact['osfamily']` to structured facts like `$fact['os']['familiy']`, we are getting issues claiming that module does not work. This is because distribs ship not a facter version that handle structured facts .

And if you add that `metedata.json` does not permit to add a requirement about Facter, as module developer IMO being able to split elements separately will not be paradise at all.
